<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:250%;text-align:center;border-radius:10px 10px; padding:10px;">BigBasket Product Recommendor</p>
<p style="background-color:#ed3237;color:#fff;font-family:newtimeroman;font-size:180%;text-align:center;border-radius:10px 10px; padding:10px;">A Beginner Friendly Guide to Creating Recommendation Systems</p>

<div style="text-align:center;">
<img src="https://m.media-amazon.com/images/S/aplus-media/sota/46b115bd-5ab5-4a74-b6b6-bb7cb4973d80.__CR0,0,600,180_PT0_SX600_V1___.png">
</div>

---
<a id='section-0'></a>
<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:120%;text-align:center;border-radius:10px 10px; padding:10px;">Table Of Contents</p>
<ol>
    <li><a href="#section-1" style="color:#5a90a7; font-size:120%;">Introduction</a></li>
    <li><a href="#section-2" style="color:#5a90a7;font-size:120%;">Importing Necessary Libraries</a></li>
    <li><a href="#section-3" style="color:#5a90a7;font-size:120%;">Data Loading and Cleaning</a></li>
    <li><a href="#section-4" style="color:#5a90a7;font-size:120%;">Exploratory Data Analysis</a></li>
    <li><a href="#section-5" style="color:#5a90a7;font-size:120%;">Demographic Filter Recommendor</a></li>
    <li><a href="#section-6" style="color:#5a90a7;font-size:120%;">Content Based Recommendor</a></li>
    <li><a href="#section-99" style="color:#5a90a7;font-size:120%;">Thank You</a></li>
</ol>

---

<a id='section-1'></a>
<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:200%;text-align:center;border-radius:10px 10px; padding:10px;">Introduction</p>

<p style="text-align:center;">
<a href="#section-0" style="background-color:#373333;color:#f1f1f1;font-family:newtimeroman;text-align:center;border-radius:10px 10px; padding:10px;">Back To Top</a>
</p>

<p style="color:#373435; font-size:16px;font-weight:400;">Have you ever noticed how when you seach for a product on Amazon, similar products are shown below it. An example of this is when you search for self-help books, Amazon shows other self-help books that might interest you. This is the job of a recommendation system.
<br><br>
Recommendation Systems are a type of information filtering systems as they improve the quality of search results and provides items that are more relevant to the search item or are realted to the search history of the user. Recommendation system finds similarity between the product a user clicks on and the other products and then recommends if their is some sustainable similarity.
<br><br>
Youtube recommends videos and even decides which video to play next on autoplay. Facebook and Instagram recommend friends and pages to folow. Instagram also recommends reels that you might be interested in. Netflix recommends movies based on user taste and genre selection, it even recommends thumbnails of movies based on user data. Amazon, BigBasket and other E-Commerce websites recommend ads and products based on the items we buy or search for.
<br><br>
In this notebook, we will be building a product recommendor with detailed explanations.
<br><br>
    <b>So Let's Start!!!</b>

There are basically three types of recommender systems:-
<ol style="color:#373435; font-size:16px;font-weight:400;">
    <li style="color:#373435; font-size:16px;font-weight:400;"><b>Demographic Filtering-</b> They offer generalized recommendations to every user, based on movie popularity and/or genre. The System recommends the same movies to users with similar demographic features. Since each user is different , this approach is considered to be too simple. The basic idea behind this system is that movies that are more popular and critically acclaimed will have a higher probability of being liked by the average audience.
    </li>
    <br>
<li style="color:#373435; font-size:16px;font-weight:400;"><b>Content Based Filtering-</b> They suggest similar items based on a particular item. This system uses item metadata, such as genre, director, description, actors, etc. for movies, to make these recommendations. The general idea behind these recommender systems is that if a person liked a particular item, he or she will also like an item that is similar to it.
    </li>
    <br>
<li style="color:#373435; font-size:16px;font-weight:400;"><b>Collaborative Filtering-</b> This system matches persons with similar interests and provides recommendations based on this matching. Collaborative filters do not require item metadata like its content-based counterparts.
    </li>
</ol>
</p>
<p style="color:#373435; font-size:16px;font-weight:400;">
Here is the deployment link of the recommendation system: <a href="https://vermaayush680-bigbasket-product-recommendation-s-mainapp-bp8r67.streamlitapp.com/" target="_blank" style="color:blue; font-size:18px;font-weight:500;">Deployment Link</a>
Will be creating a separate deployment notebook to teach it.
</p>

<a id='section-2'></a>
<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:200%;text-align:center;border-radius:10px 10px; padding:10px;">Importing Necessary Libraries</p>

<p style="text-align:center;">
<a href="#section-0" style="background-color:#373333;color:#f1f1f1;font-family:newtimeroman;text-align:center;border-radius:10px 10px; padding:10px;">Back To Top</a>
</p>

In [1]:
#Basic Libraries
import numpy as np
import pandas as pd

#Visualization Libraries
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

#Text Handling Libraries
import re
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity

<a id='section-3'></a>
<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:200%;text-align:center;border-radius:10px 10px; padding:10px;">Data Loading and Cleaning</p>

<p style="text-align:center;">
<a href="#section-0" style="background-color:#373333;color:#f1f1f1;font-family:newtimeroman;text-align:center;border-radius:10px 10px; padding:10px;">Back To Top</a>
</p>

In [2]:
df = pd.read_csv('../input/bigbasket-entire-product-list-28k-datapoints/BigBasket Products.csv',index_col='index')

In [3]:
df.head()

product  \
index                                                      
1                 Garlic Oil - Vegetarian Capsule 500 mg   
2                                  Water Bottle - Orange   
3                         Brass Angle Deep - Plain, No.2   
4      Cereal Flip Lid Container/Storage Jar - Assort...   
5                     Creme Soft Soap - For Hands & Body   

                     category           sub_category              brand  \
index                                                                     
1            Beauty & Hygiene              Hair Care  Sri Sri Ayurveda    
2      Kitchen, Garden & Pets  Storage & Accessories         Mastercook   
3        Cleaning & Household            Pooja Needs                Trm   
4        Cleaning & Household   Bins & Bathroom Ware             Nakoda   
5            Beauty & Hygiene       Bath & Hand Wash              Nivea   

       sale_price  market_price                      type  rating  \
index                                                               
1           220.0         220.0          Hair Oil & Serum     4.1   
2           180.0         180.0    Water & Fridge Bottles     2.3   
3           119.0         250.0           Lamp & Lamp Oil     3.4   
4           149.0         176.0  Laundry, Storage Baskets     3.7   
5           162.0         162.0      Bathing Bars & Soaps     4.4   

                                             description  
index                                                     
1      This Product contains Garlic Oil that is known...  
2      Each product is microwave safe (without lid), ...  
3      A perfect gift for all occasions, be it your m...  
4      Multipurpose container with an attractive desi...  
5      Nivea Creme Soft Soap gives your skin the best...

In [4]:
df.shape

(27555, 9)

In [5]:
df.isnull().sum()

product            1
category           0
sub_category       0
brand              1
sale_price         0
market_price       0
type               0
rating          8626
description      115
dtype: int64

In [6]:
print('Percentage Null Data In Each Column')
print('-'*30)
for col in df.columns:
    null_count = df[col].isnull().sum()
    total_count = df.shape[0]
    print("{} : {:.2f}".format(col,null_count/total_count * 100))
    

Percentage Null Data In Each Column
------------------------------
product : 0.00
category : 0.00
sub_category : 0.00
brand : 0.00
sale_price : 0.00
market_price : 0.00
type : 0.00
rating : 31.30
description : 0.42


In [7]:
print('Total Null Data')
null_count = df.isnull().sum().sum()
total_count = np.product(df.shape)
print("{:.2f}".format(null_count/total_count * 100))

Total Null Data
3.53


<p style="color:#373435; font-size:16px;font-weight:400;">So overall 3% data is missing but 31% of ratings are missing.
Since we are going to create a recommender system, let's drop the null values as their will still be over 69% data for recommendation purposes which is enough for us.</p>

In [8]:
df = df.dropna()

In [9]:
df.isnull().sum()

product         0
category        0
sub_category    0
brand           0
sale_price      0
market_price    0
type            0
rating          0
description     0
dtype: int64

In [10]:
df.shape

(18840, 9)

In [11]:
# df.to_csv('data_cleaned.csv')

<p style="color:#373435; font-size:16px;font-weight:400;">So even after dropping null data, 18000+ products are available for recommendation.<br>
Let's recommend now!!</p>

<a id='section-4'></a>
<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:200%;text-align:center;border-radius:10px 10px; padding:10px;">Exploratory Data Analysis</p>

<p style="text-align:center;">
<a href="#section-0" style="background-color:#373333;color:#f1f1f1;font-family:newtimeroman;text-align:center;border-radius:10px 10px; padding:10px;">Back To Top</a>
</p>

In [12]:
df.head()

product  \
index                                                      
1                 Garlic Oil - Vegetarian Capsule 500 mg   
2                                  Water Bottle - Orange   
3                         Brass Angle Deep - Plain, No.2   
4      Cereal Flip Lid Container/Storage Jar - Assort...   
5                     Creme Soft Soap - For Hands & Body   

                     category           sub_category              brand  \
index                                                                     
1            Beauty & Hygiene              Hair Care  Sri Sri Ayurveda    
2      Kitchen, Garden & Pets  Storage & Accessories         Mastercook   
3        Cleaning & Household            Pooja Needs                Trm   
4        Cleaning & Household   Bins & Bathroom Ware             Nakoda   
5            Beauty & Hygiene       Bath & Hand Wash              Nivea   

       sale_price  market_price                      type  rating  \
index                                                               
1           220.0         220.0          Hair Oil & Serum     4.1   
2           180.0         180.0    Water & Fridge Bottles     2.3   
3           119.0         250.0           Lamp & Lamp Oil     3.4   
4           149.0         176.0  Laundry, Storage Baskets     3.7   
5           162.0         162.0      Bathing Bars & Soaps     4.4   

                                             description  
index                                                     
1      This Product contains Garlic Oil that is known...  
2      Each product is microwave safe (without lid), ...  
3      A perfect gift for all occasions, be it your m...  
4      Multipurpose container with an attractive desi...  
5      Nivea Creme Soft Soap gives your skin the best...

In [13]:
counts = df['category'].value_counts()

counts_df = pd.DataFrame({'Category':counts.index,'Counts':counts.values})

In [14]:
px.bar(data_frame=counts_df,
 x='Category',
 y='Counts',
 color='Counts',
 color_continuous_scale='blues',
 text_auto=True,
 title=f'Count of Items in Each Category')


In [15]:
counts = df['sub_category'].value_counts()

counts_df_1 = pd.DataFrame({'Category':counts.index,'Counts':counts.values})[:10]

In [16]:
px.bar(data_frame=counts_df_1,
 x='Category',
 y='Counts',
 color='Counts',
 color_continuous_scale='blues',
 text_auto=True,
 title=f'Top 10 Bought Sub_Categories')


In [17]:
counts = df['brand'].value_counts()

counts_df_brand = pd.DataFrame({'Brand Name':counts.index,'Counts':counts.values})[:10]

In [18]:
px.bar(data_frame=counts_df_brand,
 x='Brand Name',
 y='Counts',
 color='Counts',
 color_continuous_scale='blues',
 text_auto=True,
 title=f'Top 10 Brand Items based on Item Counts')

In [19]:
counts = df['type'].value_counts()

counts_df_type = pd.DataFrame({'Type':counts.index,'Counts':counts.values})[:10]

In [20]:
px.bar(data_frame=counts_df_type,
 x='Type',
 y='Counts',
 color='Counts',
 color_continuous_scale='blues',
 text_auto=True,
 title=f'Top 10 Types of Products based on Item Counts')

<a id='section-5'></a>
<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:200%;text-align:center;border-radius:10px 10px; padding:10px;">Demographic Filter Recommendor</p>

<p style="text-align:center;">
<a href="#section-0" style="background-color:#373333;color:#f1f1f1;font-family:newtimeroman;text-align:center;border-radius:10px 10px; padding:10px;">Back To Top</a>
</p>

<p style="color:#373435; font-size:16px;font-weight:400;">Demographic Filtering is like recommeding items based on a feature. Like the top 10 rated items or the top 10 items in a particular category.</p>

In [21]:
def sort_recommendor(col='rating',sort_type = False):
    """
    A recommendor based on sorting products on the column passed.
    Arguments to be passed:
    
    col: The Feature to be used for recommendation.
    sort_type: True for Ascending Order
    """
    rated_recommend = df.copy()
    if rated_recommend[col].dtype == 'O':
        col='rating'
    rated_recommend = rated_recommend.sort_values(by=col,ascending = sort_type)
    return rated_recommend[['product','brand','sale_price','rating']].head(10)

In [22]:
help(sort_recommendor)

Help on function sort_recommendor in module __main__:

sort_recommendor(col='rating', sort_type=False)
    A recommendor based on sorting products on the column passed.
    Arguments to be passed:
    
    col: The Feature to be used for recommendation.
    sort_type: True for Ascending Order



In [23]:
sort_recommendor(col='sale_price',sort_type=True)

product         brand  sale_price  rating
index                                                                        
21313                                 Serum         Livon         3.0     2.5
18291                Sugar Coated Chocolate  Cadbury Gems         5.0     4.2
21229                        Dish Shine Bar           Exo         5.0     4.2
14539          Cadbury Perk - Chocolate Bar       Cadbury         5.0     4.2
19539                Layer Cake - Chocolate       Winkies         5.0     4.2
2979   Sugar Free Chewing Gum - Mixed Fruit         Orbit         5.0     4.2
15927               Dreams Cup Cake - Choco         Elite         5.0     3.9
6015                Good Day Butter Cookies     Britannia         5.0     4.1
27414                   Layer Cake - Orange       Winkies         5.0     4.1
11307        Happy Happy Choco-Chip Cookies         Parle         5.0     4.2

<p style="color:#373435; font-size:16px;font-weight:400;">Notice that our top product has rating of 2.5 which is quite bad so let's filter down by setting a threshold rating.</p>

In [24]:
C= df['rating'].mean()
C

3.9430626326963902

<p style="color:#373435; font-size:16px;font-weight:400;">So the average rating of products is 3.94
Let's use 3.5 as the threshold.</p>

In [25]:
def sort_recommendor(col='rating',sort_type = False):
    """
    A recommendor based on sorting products on the column passed.
    Arguments to be passed:
    
    col: The Feature to be used for recommendation.
    sort_type: True for Ascending Order
    """
    rated_recommend = df.copy().loc[df['rating'] >= 3.5]
    if rated_recommend[col].dtype == 'O':
        col='rating'
    rated_recommend = rated_recommend.sort_values(by=col,ascending = sort_type)
    return rated_recommend[['product','brand','sale_price','rating']].head(10)

In [26]:
sort_recommendor(col='sale_price',sort_type=True)

product      brand  sale_price  \
index                                                                       
2762   Orbit Sugar-Free Chewing Gum - Lemon & Lime   Wrigleys         5.0   
3446                 Marie Light Biscuits - Active   Sunfeast         5.0   
14604                      50-50 Timepass Biscuits  Britannia         5.0   
17641                  Hand Wash - Moisture Shield     Savlon         5.0   
27491               50-50 Timepass Salted Biscuits  Britannia         5.0   
26585                Polo - The Mint With The Hole    Nestle          5.0   
2979          Sugar Free Chewing Gum - Mixed Fruit      Orbit         5.0   
19539                       Layer Cake - Chocolate    Winkies         5.0   
19203                Bounce Biscuits - Choco Creme   Sunfeast         5.0   
14539                 Cadbury Perk - Chocolate Bar    Cadbury         5.0   

       rating  
index          
2762      4.2  
3446      4.5  
14604     3.9  
17641     4.4  
27491     4.2  
26585     4.4  
2979      4.2  
19539     4.2  
19203     4.2  
14539     4.2

<p style="color:#373435; font-size:16px;font-weight:400;">Notice that the 2.5 rated product is not recommended now!!
This was our first recommendor.
<br>
Quite easy yet effective and used a lot !!</p>

<a id='section-6'></a>
<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:200%;text-align:center;border-radius:10px 10px; padding:10px;">Content Based Recommendor</p>

<p style="text-align:center;">
<a href="#section-0" style="background-color:#373333;color:#f1f1f1;font-family:newtimeroman;text-align:center;border-radius:10px 10px; padding:10px;">Back To Top</a>
</p>

In [27]:
df.head()

product  \
index                                                      
1                 Garlic Oil - Vegetarian Capsule 500 mg   
2                                  Water Bottle - Orange   
3                         Brass Angle Deep - Plain, No.2   
4      Cereal Flip Lid Container/Storage Jar - Assort...   
5                     Creme Soft Soap - For Hands & Body   

                     category           sub_category              brand  \
index                                                                     
1            Beauty & Hygiene              Hair Care  Sri Sri Ayurveda    
2      Kitchen, Garden & Pets  Storage & Accessories         Mastercook   
3        Cleaning & Household            Pooja Needs                Trm   
4        Cleaning & Household   Bins & Bathroom Ware             Nakoda   
5            Beauty & Hygiene       Bath & Hand Wash              Nivea   

       sale_price  market_price                      type  rating  \
index                                                               
1           220.0         220.0          Hair Oil & Serum     4.1   
2           180.0         180.0    Water & Fridge Bottles     2.3   
3           119.0         250.0           Lamp & Lamp Oil     3.4   
4           149.0         176.0  Laundry, Storage Baskets     3.7   
5           162.0         162.0      Bathing Bars & Soaps     4.4   

                                             description  
index                                                     
1      This Product contains Garlic Oil that is known...  
2      Each product is microwave safe (without lid), ...  
3      A perfect gift for all occasions, be it your m...  
4      Multipurpose container with an attractive desi...  
5      Nivea Creme Soft Soap gives your skin the best...

<p style="color:#373435; font-size:16px;font-weight:400;">Let's try using other features such as Category, Sub Category, Brand, Type and Description for much better Recommendation.<br>
We will be using NLP here to extract useful info from the features especially Description so let's understand TF-IDF before using it.</p>

<p style="color:#373435; font-size:16px;font-weight:400;">
    <b>TF-IDF</b> stands for <b>term frequency-inverse document frequency</b>.
</p>
<p style="color:#373435; font-size:18px;font-weight:500;">
    <b>What is TF(Term Frequency):</b>
</p>
<p style="color:#373435; font-size:16px;font-weight:400;">
Term frequency works by looking at the frequency of a particular term you are concerned with relative to the document. There are multiple measures, or ways, of defining frequency: Number of times the word appears in a document (raw count).
<br><br>
Term frequency adjusted for the length of the document (raw count of occurences divided by number of words in the document).
Logarithmically scaled frequency (e.g. log(1 + raw count)).
Boolean frequency (e.g. 1 if the term occurs, or 0 if the term does not occur, in the document).
</p>

<p style="color:#373435; font-size:18px;font-weight:500;">
    <b>What is IDF (inverse document frequency)?</b>
<p>
<p style="color:#373435; font-size:16px;font-weight:400;">
Inverse document frequency looks at how common (or uncommon) a word is amongst the corpus. IDF is calculated as follows where t is the term (word) we are looking to measure the commonness of and N is the number of documents (d) in the corpus (D).. 
<br>
The denominator is simply the number of documents in which the term, t, appears in. 
    
<div style="text-align:center;">
<img src="https://ecm.capitalone.com/WCM/tech/tf-idf-1.png">
</div>
    
<br>
<div style="text-align:center;">
    <p style="color:#373435; font-size:20px;font-weight:600;text-align:center">
    <b>TF-IDF at work</b>
    </p>
<img src="https://ecm.capitalone.com/WCM/tech/tf-idf-4.png"/>
</div>
<div style="text-align:center;">
    <a href="https://www.capitalone.com/tech/machine-learning/understanding-tf-idf/" target="_blank" style="color:blue; font-size:20px;font-weight:600;text-align:center">Source Credits</a>
</div>
    

In [28]:
tfidf = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf.fit_transform(df['description'])
tfidf_matrix.shape

(18840, 23342)

<p style="color:#373435; font-size:16px;font-weight:400;">
Now to compute the similarity score, let's use Linear_Kernel.
    <br>
    Linear Kernel which Calculates the Dot Product of the tfidf_matrix and returns an aggreate value depicting the Similarity score.
</p>

In [29]:
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)
cosine_sim

array([[1.        , 0.01632718, 0.00999603, ..., 0.01056047, 0.01133156,
        0.        ],
       [0.01632718, 1.        , 0.00719713, ..., 0.        , 0.        ,
        0.        ],
       [0.00999603, 0.00719713, 1.        , ..., 0.00635776, 0.        ,
        0.        ],
       ...,
       [0.01056047, 0.        , 0.00635776, ..., 1.        , 0.        ,
        0.        ],
       [0.01133156, 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

<p style="color:#373435; font-size:16px;font-weight:400;">
So we will be recommending items based on similarity score.<br>
But our problem is that we will be getting back the similarity scores so we will be sorting the scores.<br>
Now we need a reverse-map to get the title and that is what <b>indices</b> is for.
</p>

In [30]:
indices = pd.Series(df.index, index=df['product']).drop_duplicates()

def get_recommendations_1(title, cosine_sim=cosine_sim):
    
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:11]
    movie_indices = [i[0] for i in sim_scores]
    return df['product'].iloc[movie_indices]

In [31]:
get_recommendations_1('Water Bottle - Orange')

index
1677         Brass Nanda Stand Goblets - No.1
2162         Brass Kachua Stand Deepam - No.1
2756     Brass Angle Deep Stand - Plain, No.2
5400       Brass Lakshmi Deepam - Plain, No.2
6520                Brass Kuber Deepam - No.1
10504               Brass Kuber Deepam - No.2
11226    Brass Angle Deep Stand - Plain, No.3
11504    Brass Angle Deep Stand - Plain, No.1
12699        Brass Kachua Stand Deepam - No.2
18572               Brass Kuber Deepam - No.3
Name: product, dtype: object

In [32]:
get_recommendations_1('Cadbury Perk - Chocolate Bar')

index
27049                        Pickle - Mixed
6601                  Pickle - Kaduku Mango
17934                Pickle - Mix Vegetable
27105                        Pickle - Prawn
3962                  Pickle - Tender Mango
16875             Olive Oil - Carrot Pickle
3444                     Pickle - Cut Mango
17237      Andhra Special Red Chilli Pickle
27234    Pickle - Lime (South Indian Style)
4955                    Pickle - Gooseberry
Name: product, dtype: object

<p style="color:#373435; font-size:16px;font-weight:400;">
Our search was chocolate yet we got Cashews and Nuts recommended. <br>
We need to optimize this based on category, sub_category and brand.
</p>

In [33]:
df2 = df.copy()

In [34]:
df2.head()

product  \
index                                                      
1                 Garlic Oil - Vegetarian Capsule 500 mg   
2                                  Water Bottle - Orange   
3                         Brass Angle Deep - Plain, No.2   
4      Cereal Flip Lid Container/Storage Jar - Assort...   
5                     Creme Soft Soap - For Hands & Body   

                     category           sub_category              brand  \
index                                                                     
1            Beauty & Hygiene              Hair Care  Sri Sri Ayurveda    
2      Kitchen, Garden & Pets  Storage & Accessories         Mastercook   
3        Cleaning & Household            Pooja Needs                Trm   
4        Cleaning & Household   Bins & Bathroom Ware             Nakoda   
5            Beauty & Hygiene       Bath & Hand Wash              Nivea   

       sale_price  market_price                      type  rating  \
index                                                               
1           220.0         220.0          Hair Oil & Serum     4.1   
2           180.0         180.0    Water & Fridge Bottles     2.3   
3           119.0         250.0           Lamp & Lamp Oil     3.4   
4           149.0         176.0  Laundry, Storage Baskets     3.7   
5           162.0         162.0      Bathing Bars & Soaps     4.4   

                                             description  
index                                                     
1      This Product contains Garlic Oil that is known...  
2      Each product is microwave safe (without lid), ...  
3      A perfect gift for all occasions, be it your m...  
4      Multipurpose container with an attractive desi...  
5      Nivea Creme Soft Soap gives your skin the best...

In [35]:
df2.shape

(18840, 9)

<p style="color:#373435; font-size:16px;font-weight:400;">
Notice that a product can be in multiple catergories and sub_categories and they are separated with a &.<br>
Let's split them into a list for futher processes.
</p>

In [36]:
rmv_spc = lambda a:a.strip()
get_list = lambda a:list(map(rmv_spc,re.split('& |, |\*|\n', a)))

In [37]:
get_list('A & B, C')

['A', 'B', 'C']

In [38]:
for col in ['category', 'sub_category', 'type']:
    df2[col] = df2[col].apply(get_list)

In [39]:
df2.head()

product  \
index                                                      
1                 Garlic Oil - Vegetarian Capsule 500 mg   
2                                  Water Bottle - Orange   
3                         Brass Angle Deep - Plain, No.2   
4      Cereal Flip Lid Container/Storage Jar - Assort...   
5                     Creme Soft Soap - For Hands & Body   

                      category            sub_category              brand  \
index                                                                       
1            [Beauty, Hygiene]             [Hair Care]  Sri Sri Ayurveda    
2      [Kitchen, Garden, Pets]  [Storage, Accessories]         Mastercook   
3        [Cleaning, Household]           [Pooja Needs]                Trm   
4        [Cleaning, Household]   [Bins, Bathroom Ware]             Nakoda   
5            [Beauty, Hygiene]       [Bath, Hand Wash]              Nivea   

       sale_price  market_price                        type  rating  \
index                                                                 
1           220.0         220.0           [Hair Oil, Serum]     4.1   
2           180.0         180.0     [Water, Fridge Bottles]     2.3   
3           119.0         250.0            [Lamp, Lamp Oil]     3.4   
4           149.0         176.0  [Laundry, Storage Baskets]     3.7   
5           162.0         162.0       [Bathing Bars, Soaps]     4.4   

                                             description  
index                                                     
1      This Product contains Garlic Oil that is known...  
2      Each product is microwave safe (without lid), ...  
3      A perfect gift for all occasions, be it your m...  
4      Multipurpose container with an attractive desi...  
5      Nivea Creme Soft Soap gives your skin the best...

<p style="color:#373435; font-size:16px;font-weight:400;">
To avoid duplicacy, we will be converting everything to lowercase and also removing spaces between words. <br>
    This will ensure that our recommendor doesn't consider <b>Chocolate</b> of <b>Cholocate IceCream</b> and <b>Chocolate Bar</b> as the same.
</p>

In [40]:
def cleaner(x):
    if isinstance(x, list):
        return [str.lower(i.replace(" ", "")) for i in x]
    else:
        if isinstance(x, str):
            return str.lower(x.replace(" ", ""))
        else:
            return ''

In [41]:
for col in ['category', 'sub_category', 'type','brand']:
    df2[col] = df2[col].apply(cleaner)

In [42]:
df2.head()

product  \
index                                                      
1                 Garlic Oil - Vegetarian Capsule 500 mg   
2                                  Water Bottle - Orange   
3                         Brass Angle Deep - Plain, No.2   
4      Cereal Flip Lid Container/Storage Jar - Assort...   
5                     Creme Soft Soap - For Hands & Body   

                      category            sub_category           brand  \
index                                                                    
1            [beauty, hygiene]              [haircare]  srisriayurveda   
2      [kitchen, garden, pets]  [storage, accessories]      mastercook   
3        [cleaning, household]            [poojaneeds]             trm   
4        [cleaning, household]    [bins, bathroomware]          nakoda   
5            [beauty, hygiene]        [bath, handwash]           nivea   

       sale_price  market_price                       type  rating  \
index                                                                
1           220.0         220.0           [hairoil, serum]     4.1   
2           180.0         180.0     [water, fridgebottles]     2.3   
3           119.0         250.0            [lamp, lampoil]     3.4   
4           149.0         176.0  [laundry, storagebaskets]     3.7   
5           162.0         162.0       [bathingbars, soaps]     4.4   

                                             description  
index                                                     
1      This Product contains Garlic Oil that is known...  
2      Each product is microwave safe (without lid), ...  
3      A perfect gift for all occasions, be it your m...  
4      Multipurpose container with an attractive desi...  
5      Nivea Creme Soft Soap gives your skin the best...

<p style="color:#373435; font-size:16px;font-weight:400;">
We will now be joining the values of category, sub_category, type and brand
</p>

In [43]:
def couple(x):
    return ' '.join(x['category']) + ' ' + ' '.join(x['sub_category']) + ' '+x['brand']+' ' +' '.join( x['type'])
df2['soup'] = df2.apply(couple, axis=1)

In [44]:
df2['soup'].head()

index
1    beauty hygiene haircare srisriayurveda hairoil...
2    kitchen garden pets storage accessories master...
3       cleaning household poojaneeds trm lamp lampoil
4    cleaning household bins bathroomware nakoda la...
5    beauty hygiene bath handwash nivea bathingbars...
Name: soup, dtype: object

<p style="color:#373435; font-size:16px;font-weight:400;">
We need to Count the String Vectors and then compute the Cosine Similarity Score.
</p>

In [45]:
df2.head()

product  \
index                                                      
1                 Garlic Oil - Vegetarian Capsule 500 mg   
2                                  Water Bottle - Orange   
3                         Brass Angle Deep - Plain, No.2   
4      Cereal Flip Lid Container/Storage Jar - Assort...   
5                     Creme Soft Soap - For Hands & Body   

                      category            sub_category           brand  \
index                                                                    
1            [beauty, hygiene]              [haircare]  srisriayurveda   
2      [kitchen, garden, pets]  [storage, accessories]      mastercook   
3        [cleaning, household]            [poojaneeds]             trm   
4        [cleaning, household]    [bins, bathroomware]          nakoda   
5            [beauty, hygiene]        [bath, handwash]           nivea   

       sale_price  market_price                       type  rating  \
index                                                                
1           220.0         220.0           [hairoil, serum]     4.1   
2           180.0         180.0     [water, fridgebottles]     2.3   
3           119.0         250.0            [lamp, lampoil]     3.4   
4           149.0         176.0  [laundry, storagebaskets]     3.7   
5           162.0         162.0       [bathingbars, soaps]     4.4   

                                             description  \
index                                                      
1      This Product contains Garlic Oil that is known...   
2      Each product is microwave safe (without lid), ...   
3      A perfect gift for all occasions, be it your m...   
4      Multipurpose container with an attractive desi...   
5      Nivea Creme Soft Soap gives your skin the best...   

                                                    soup  
index                                                     
1      beauty hygiene haircare srisriayurveda hairoil...  
2      kitchen garden pets storage accessories master...  
3         cleaning household poojaneeds trm lamp lampoil  
4      cleaning household bins bathroomware nakoda la...  
5      beauty hygiene bath handwash nivea bathingbars...

In [46]:
df2.to_csv('data_cleaned_1.csv')

In [47]:
count = CountVectorizer(stop_words='english')
count_matrix = count.fit_transform(df2['soup'])

<p style="color:#373435; font-size:16px;font-weight:400;">
We need to Count the String Vectors and then compute the Cosine Similarity Score.
</p>

In [48]:
cosine_sim2 = cosine_similarity(count_matrix, count_matrix)
cosine_sim2

array([[1.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.27216553],
       [0.        , 1.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.27216553, 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

<a href="https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html#sklearn.metrics.pairwise.cosine_similarity" target="_blank" style="color:blue; font-size:16px;font-weight:400;">
Cosine Similarity Documentation
</a>

In [49]:
df2 = df2.reset_index()
indices = pd.Series(df2.index, index=df2['product'])

In [50]:

def get_recommendations_2(title, cosine_sim=cosine_sim):
    idx = indices[title]

    sim_scores = list(enumerate(cosine_sim[idx]))

    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    sim_scores = sim_scores[1:11]

    movie_indices = [i[0] for i in sim_scores]

    return df2['product'].iloc[movie_indices]

<p style="color:#373435; font-size:16px;font-weight:400;">
Comparing Old and New Recommedations
</p>

In [51]:
old_rec = get_recommendations_1('Water Bottle - Orange').values
new_rec = get_recommendations_2('Water Bottle - Orange', cosine_sim2).values

pd.DataFrame({'Old Recommendor': old_rec,'New Recommendor':new_rec})

Old Recommendor  \
0  Rectangular Plastic Container - With Lid, Mult...   
1                             Jar - With Lid, Yellow   
2   Round & Flat Storage Container - With lid, Green   
3  Premium Rectangular Plastic Container With Lid...   
4  Premium Round Plastic Container With Lid - Yellow   
5  Premium Rectangular Plastic Container With Lid...   
6  Premium Round & Flat Storage Container With Li...   
7    Premium Round Plastic Container With Lid - Blue   
8  Premium Round Plastic Container With Lid - Mul...   
9    Premium Round Plastic Container With Lid - Pink   

                                     New Recommendor  
0        Glass Water Bottle - Aquaria Organic Purple  
1  Glass Water Bottle With Round Base - Transpare...  
2                H2O Unbreakable Water Bottle - Pink  
3                            Water Bottle H2O Purple  
4               H2O Unbreakable Water Bottle - Green  
5   Regel Tritan Plastic Sports Water Bottle - Black  
6            Apsara 1 Water Bottle - Assorted Colour  
7  Glass Water Bottle With Round Base - Yellow, B...  
8  Trendy Stainless Steel Bottle With Steel Cap -...  
9  Penta Plastic Pet Water Bottle - Violet, Wide ...

In [52]:
old_rec = get_recommendations_1('Cadbury Perk - Chocolate Bar').values
new_rec = get_recommendations_2('Cadbury Perk - Chocolate Bar', cosine_sim2).values

pd.DataFrame({'Old Recommendor': old_rec,'New Recommendor':new_rec})

Old Recommendor  \
0              Cadbury Perk - Chocolate Bar   
1                Choco Stick - Hexagon Pack   
2  Luvit Chocwich White Home Delights 187 g   
3        Luvit Chocwich Home Delights 187 g   
4         Wafer Biscuits - Chocolate Flavor   
5             Drinking Chocolate - Original   
6             Drinking Chocolate - Original   
7                  Biscuit - Bourbon Creams   
8                Wafers With Hazelnut Cream   
9                   Choco Stick - Chocolate   

                                    New Recommendor  
0                            Nutties Chocolate Pack  
1                              5 Star Chocolate Bar  
2          Dairy Milk Silk - Hazelnut Chocolate Bar  
3  Perk - Chocolate, Home Treats, 175.5 g, 27 Units  
4                           Dark Milk Chocolate Bar  
5            Dairy Milk Silk Mousse - Chocolate Bar  
6                           Dark Milk Chocolate Bar  
7                              Chocolate Bar - Fuse  
8                             Choclairs Gold Coffee  
9       5 Star Chocolate Home Pack, 200 g, 20 units

<p style="color:#373435; font-size:16px;font-weight:400;">
Our new recommendation are much better compared to the old ones.
</p>

<a id='section-99'></a>
<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:250%;text-align:center;border-radius:10px 10px; padding:10px;">Thank You</p>
<p style="background-color:#373435;color:#fff;font-family:newtimeroman;font-size:180%;text-align:center;border-radius:10px 10px; padding:10px;">If you found it useful, do upvote it!!!</p>

<p style="text-align:center;">
<a href="#section-0" style="background-color:#373333;color:#f1f1f1;font-family:newtimeroman;text-align:center;border-radius:10px 10px; padding:10px;">Back To Top</a>
</p>

<div style="text-align:center;">
<img src="https://m.media-amazon.com/images/S/aplus-media/sota/46b115bd-5ab5-4a74-b6b6-bb7cb4973d80.__CR0,0,600,180_PT0_SX600_V1___.png">
</div>

---